File: BOJ/Silver/Silver1/18233.py
import sys

def duck(k, idx) :
    global s1, s2
    if k == P :

        if s1 <= E <= s2 :
            cal()
        return
    for i in range (N) :
        if used_lst[i] != 1 :
            if idx > i :
                continue
            used_lst[i] = 1
            s1 += lst[i][0]
            s2 += lst[i][1]
            duck(k + 1, i)
            s1 -= lst[i][0]
            s2 -= lst[i][1]
            used_lst[i] = 0

def cal() :
    global E
    global ans_lst
    for i in range (N) :
        if used_lst[i] == 1 :
            ans_lst[i] = lst[i][0]
    remain = E - sum(ans_lst)
    for i in range (N) :
        if used_lst[i] == 1 :
            x = lst[i][1]-lst[i][0]
            if remain >= x :
                ans_lst[i] = lst[i][1]
                remain -= x
            else :
                ans_lst[i] += remain
                remain = 0
            if remain == 0 :
                break
    for j in ans_lst :
        print (j, end=" ")
    exit()

# 변수 설정
global E
global ans_lst
N, P, E = map(int, sys.stdin.readline().split())
lst = []
for i in range (N) :
    tmp_lst = list(map(int, sys.stdin.readline().split()))
    lst.append(tmp_lst)
if N < P :
    print(-1)
    exit()
used_lst = [0] * N
ans_lst = [0] * N
global s1, s2
s1 = 0
s2 = 0
# 실행
duck(0, -1)
print(-1)

File: C/Silver/Silver1/18233.cpp
//#include <stdio.h>
//
//int N, P, E, s;
////int TF = 0;
//int lst[20][2], used_lst[20], ans_lst[20];
//
//void duck(int k);
//void cal(int k);
//
//int main()
//{
//	scanf("%d %d %d", &N, &P, &E);
//	for (int i = 0; i < N; i++)
//	{
//		scanf("%d %d", &lst[i][0], &lst[i][1]);
//	}
//	duck(0);
//	printf("%d", -1);
//	return 0;
//}
//
//void duck(int k)
//{
//	//if (TF) return;
//
//	if (k == P)
//	{
//		for (int i = 0; i < N; i++)
//		{
//			ans_lst[i] = 0;
//		}
//		cal(0);
//		return;
//	}
//
//	for (int i = 0; i < N; i++)
//	{
//		if (!used_lst[i])
//		{
//			used_lst[i] = 1;
//			duck(k + 1);
//			used_lst[i] = 0;
//		}
//	}
//}
//
//void cal(int k)
//{
//	//if (TF) return;
//
//	if (k == P)
//	{
//		// s = sum(ans_lst)
//		s = 0;
//		for (int i = 0; i < N; i++)
//		{
//			s += ans_lst[i];
//		}
//		//printf("%d ", s);
//		if (s == E)
//		{
//			for (int i = 0; i < N; i++)
//			{
//				printf("%d ", ans_lst[i]);
//			}
//			exit(0);
//		}
//	}
//
//	for (int i = 0; i < N; i++)
//	{
//		if (used_lst[i])
//		{
//			for (int j = lst[i][0]; j <= lst[i][1]; j++)
//			{
//				used_lst[i] = 0;
//				ans_lst[i] = j;
//				cal(k + 1);
//				//if (TF) return;
//				used_lst[i] = 1;
//			}
//		}
//	}
//}
